### PR TITLE
Mapswipe support

### PR DESCRIPTION
**Pull request by AGMETEOR**
## Description
This feature adds the ability to activate layer swiping for open layers. The user opens the TOC and selects a layer, then activates swiping from the TOC toolbar.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**

#5900 
#5891 
#5892 

**What is the new behavior?**
Activating tool
----
activating

Vertical Swiping
----
swiping

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/components/swipe/EffectSupport.jsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
+/*
+* Copyright 2020, GeoSolutions Sas.
+* All rights reserved.
+*
+* This source code is licensed under the BSD-style license found in the
```
Comment (offtherailz) [inline web/client/components/swipe/EffectSupport.jsx:5]: This is the EffectSupport implementation for Openlayers ,it should go together with the other supports in `openlayers` directory.

File: web/client/actions/layers.js [hunk 2]
```diff
@@ -52,6 +54,21 @@ function hideSettings() {
     };
 }
 
+function showLayerSwipeSettings(node, nodeType) {
```
Comment (offtherailz) [inline web/client/actions/layers.js:57]: these are swipe specific actions, create a dedicated file

File: web/client/actions/layers.js [hunk 3]
```diff
@@ -303,13 +320,69 @@ function updateSettingsParams(newParams, update) {
 }
 
 module.exports = {
-    changeLayerProperties, changeLayerParams, changeGroupProperties, toggleNode, sortNode, removeNode, contextNode,
-    updateNode, moveNode, layerLoading, layerLoad, layerError, addLayer, removeLayer, showSettings, hideSettings, updateSettings, refreshLayers,
-    layersRefreshed, layersRefreshError, refreshLayerVersion, updateLayerDimension, browseData, clearLayers, selectNode, filterLayers, showLayerMetadata,
-    hideLayerMetadata, download, updateSettingsParams, addGroup,
-    CHANGE_LAYER_PROPERTIES, CHANGE_LAYER_PARAMS, CHANGE_GROUP_PROPERTIES, TOGGLE_NODE, SORT_NODE,
-    REMOVE_NODE, UPDATE_NODE, MOVE_NODE, LAYER_LOADING, LAYER_LOAD, LAYER_ERROR, ADD_LAYER, REMOVE_LAYER,
+    changeLayerProperties,
```
Comment (offtherailz) [inline web/client/actions/layers.js:323]: reset also this reformat please, when we refactor this file, we want to move to `export` syntax

File: web/client/components/map/openlayers/swipe/XYSwipeSupport.jsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                left: `${map.getProperties().size[0] / 2}px`,
+                width: "12px",
+                cursor: "col-resize"
+            }}></div>
+        </Draggable>
+    );
+};
+
+// Allow swiping in the horizontal or vertical direction
+const XYSwipeSupport = ({ map, layer, type = "cut-vertical", active }) => {
```
Comment (offtherailz) [inline web/client/components/map/openlayers/swipe/XYSwipeSupport.jsx:57]: XY is still ambiguous... mmm ... what about  "SliderSwipeSupport" vs "SpySupport" ? is this more indicative of the effective implementation?

File: web/client/epics/layers.js [hunk 5]
```diff
@@ -148,8 +157,25 @@ const updateSettingsParamsEpic = (action$, store) =>
             );
         });
 
+/**
+ * Ensures that swipeSettings active is changed back to false when a layer is deselected in TOC
+ * @memberof epics.layers
+ * @param {external:Observable} action$ manages `SELECT_NODE`
+ * @return {external:Observable}
+ */
+const resetLayerSwipeSettingsEpic = (action$, store) =>
```
Comment (offtherailz) [inline web/client/epics/layers.js:166]: Also this should be moved to a new `swipe.js` epics file

File: web/client/plugins/Swipe.jsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import React from 'react';
+import { connect } from 'react-redux';
+import { createSelector } from 'reselect';
+
+import { getSelectedLayer, layerSwipeSettingsSelector } from '../selectors/layers';
+
+import { createPlugin } from '../utils/PluginsUtils';
+
+import SwipeSettings from './SwipeSettings';
+import XYSwipeSupport from '../components/map/openlayers/swipe/XYSwipeSupport';
```
Comment (offtherailz) [inline web/client/plugins/Swipe.jsx:18]: We will need some sort of wrapper that dynamically loads the required support for the current map type (OL only).
@allyoucanmap did you already do something in your experiments with support injected by pluguins? Here we are introducing and using the Map items injection via plugin-system

File: web/client/reducers/layers.js [hunk 7]
```diff
@@ -372,6 +374,24 @@ function layers(state = { flat: [] }, action) {
             settings: settings
         });
     }
+
+    case SHOW_LAYER_SWIPE_SETTINGS: {
+        const swipeSettings = assign({}, state.swipeSettings, {
```
Comment (offtherailz) [inline web/client/reducers/layers.js:379]: Also this should be in a swipe reducer, separated from the map layers.

rename all things (actions, actionCreators and state variables) this way:

`reducers/swipe.js`:
swipeSettings changes to the to `active: true| false`

`actions/swipe.js`:
action can be only one SET_ACTIVE = "SWIPE:SET_ACTIVE", with one parameter, `active`
actionCreators in  1 or 2 (What you prefer): or `setActive(active)` or `activate() `and `deactivate()`.

don't forget the epic.

File: web/client/components/TOC/Toolbar.jsx [hunk 8]
```diff
@@ -327,6 +330,15 @@ class Toolbar extends React.Component {
                             </Button>
                         </OverlayTrigger>
                         : null}
+                    {this.props.activateTool.activateSwipeOnLayer && (status === 'LAYER') &&
```
Comment (offtherailz) [inline web/client/components/TOC/Toolbar.jsx:333]: I think that also this can be externalized, and this is a good moment to provide this functionality. I'll prepare a patch that helps to render buttons as items in TOC toolbar, but we can do it in a separate issue.

File: web/client/plugins/Swipe.jsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+const MapSwipeSupport = connect(swipeSupportSelector, null)(Support);
+
+/**
+ * Swipe. Add to the TOC the possibility to select a layer for Swipe.
+ * @memberof plugins
+ * @requires plugins.Swipe
+ */
+export default createPlugin(
+    'Swipe',
```
Comment (offtherailz) [inline web/client/plugins/Swipe.jsx:40]: use `disablePluginIf` to avoid this showing in leaflet. See mapstore dev doc

**Review by offtherailz — CHANGES_REQUESTED** [review]
Very impressive :+1:  
Please follow my suggestion to isolate this tool more. When I apply the TOC injection, this tool can be really independent from TOC and from Map.

File: web/client/epics/__tests__/layers-test.js [hunk 10]
```diff
@@ -12,8 +12,7 @@ const configureMockStore = require('redux-mock-store').default;
 const { createEpicMiddleware, combineEpics } = require('redux-observable');
 const {
     refreshLayers, LAYERS_REFRESHED, LAYERS_REFRESH_ERROR, UPDATE_NODE,
-    updateLayerDimension, CHANGE_LAYER_PARAMS, updateSettingsParams, UPDATE_SETTINGS
-} = require('../../actions/layers');
+    updateLayerDimension, CHANGE_LAYER_PARAMS, updateSettingsParams, UPDATE_SETTINGS } = require('../../actions/layers');
```
Comment (offtherailz) [inline web/client/epics/__tests__/layers-test.js:15]: leave it unchanged
```suggestion
        updateLayerDimension, CHANGE_LAYER_PARAMS, updateSettingsParams, UPDATE_SETTINGS
} = require('../../actions/layers');
```

File: web/client/reducers/layers.js [hunk 11]
```diff
@@ -8,7 +8,8 @@
 
 var { LAYER_LOADING, LAYER_LOAD, LAYER_ERROR, CHANGE_LAYER_PARAMS, CHANGE_LAYER_PROPERTIES, CHANGE_GROUP_PROPERTIES,
     TOGGLE_NODE, SORT_NODE, REMOVE_NODE, UPDATE_NODE, MOVE_NODE, ADD_LAYER, REMOVE_LAYER, ADD_GROUP,
-    SHOW_SETTINGS, HIDE_SETTINGS, UPDATE_SETTINGS, REFRESH_LAYERS, LAYERS_REFRESH_ERROR, LAYERS_REFRESHED, CLEAR_LAYERS, SELECT_NODE, FILTER_LAYERS, SHOW_LAYER_METADATA, HIDE_LAYER_METADATA
+    SHOW_SETTINGS, HIDE_SETTINGS, UPDATE_SETTINGS, REFRESH_LAYERS, LAYERS_REFRESH_ERROR,
+    LAYERS_REFRESHED, CLEAR_LAYERS, SELECT_NODE, FILTER_LAYERS, SHOW_LAYER_METADATA, HIDE_LAYER_METADATA
```
Comment (offtherailz) [inline web/client/reducers/layers.js:12]: leave it unchanged
```suggestion
SHOW_SETTINGS, HIDE_SETTINGS, UPDATE_SETTINGS, REFRESH_LAYERS, LAYERS_REFRESH_ERROR, LAYERS_REFRESHED, CLEAR_LAYERS, SELECT_NODE, FILTER_LAYERS, SHOW_LAYER_METADATA, HIDE_LAYER_METADATA
```

File: web/client/epics/__tests__/layers-test.js [hunk 12]
```diff
@@ -12,7 +12,7 @@ const configureMockStore = require('redux-mock-store').default;
 const { createEpicMiddleware, combineEpics } = require('redux-observable');
 const {
     refreshLayers, LAYERS_REFRESHED, LAYERS_REFRESH_ERROR, UPDATE_NODE,
-    updateLayerDimension, CHANGE_LAYER_PARAMS, updateSettingsParams, UPDATE_SETTINGS
+        updateLayerDimension, CHANGE_LAYER_PARAMS, updateSettingsParams, UPDATE_SETTINGS
```
Comment (offtherailz) [inline web/client/epics/__tests__/layers-test.js:15]: leave it unchanged
```suggestion
    updateLayerDimension, CHANGE_LAYER_PARAMS, updateSettingsParams, UPDATE_SETTINGS```

File: web/client/epics/__tests__/layers-test.js [hunk 13]
```diff
@@ -12,7 +12,7 @@ const configureMockStore = require('redux-mock-store').default;
 const { createEpicMiddleware, combineEpics } = require('redux-observable');
 const {
     refreshLayers, LAYERS_REFRESHED, LAYERS_REFRESH_ERROR, UPDATE_NODE,
-    updateLayerDimension, CHANGE_LAYER_PARAMS, updateSettingsParams, UPDATE_SETTINGS
+    updateLayerDimension, CHANGE_LAYER_PARAMS, updateSettingsParams, UPDATE_SETTINGS```
```
Comment (offtherailz) [inline web/client/epics/__tests__/layers-test.js:15]: fix my typo
```suggestion
    updateLayerDimension, CHANGE_LAYER_PARAMS, updateSettingsParams, UPDATE_SETTINGS
```

File: web/client/components/TOC/Toolbar.jsx [hunk 14]
```diff
@@ -327,6 +329,15 @@ class Toolbar extends React.Component {
...
+                        <OverlayTrigger
+                            key="layerSwipe"
+                            placement="top"
+                            overlay={<Tooltip id="layer-tooltip-swipe">Layer swipe settings</Tooltip>}>
```
Comment (offtherailz) [inline web/client/components/TOC/Toolbar.jsx:336]: Please localize this string.

File: web/client/components/map/openlayers/swipe/SliderSwipeSupport.jsx [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+                top: '0px',
+                left: `${map.getProperties().size[0] / 2}px`,
+                width: "12px",
+                cursor: "col-resize"
+            }}></div>
+        </Draggable>
+    );
+};
+
+// Allow swiping in the horizontal or vertical direction
```
Comment (offtherailz) [inline web/client/components/map/openlayers/swipe/SliderSwipeSupport.jsx:56]:
```suggestion
/**
 * Implementation of Horiziontal/Vertical slider swipe for OpenLayers. 
 * @props {string} [type="cut-vertical"] the type of the swipe. Can be "cut-vertical" or "cut-horizontal"
 * @props {boolean} active activates the tool (if a layer is present)
 * @props {object} map the map object
 * @props {layer} the layer object
 */
```

File: web/client/plugins/Swipe.jsx [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+import * as epics from '../epics/swipe';
+
+import { createPlugin } from '../utils/PluginsUtils';
+
+import SwipeSettings from './SwipeSettings';
+import SliderSwipeSupport from '../components/map/openlayers/swipe/SliderSwipeSupport';
+
+const Support = ({ map, layer, active }) => {
+    return <SliderSwipeSupport map={map} layer={layer} active={active} />;
+};
```
Comment (offtherailz) [inline web/client/plugins/Swipe.jsx:25]: We have to isolate somehow this implementation in a separate component. "SwipeSupport" and wrap into a generic components that loads the OL implementation on demand.
But we can do it in a separate PR. @allyoucanmap, what do you think ? Can we do in a next separated PR ?

**Review by offtherailz — CHANGES_REQUESTED** [review]
Very Good !!
This PR is very clean, almost perfect :+1: I give you my personal compliments about this

I only asked to localize one string ("I should use "Compare" for the tooltip of the TOC button, let's ask to @tdipisa ) and document more properties of the support. 

Don't forget to isolate the implementation ( I think in a next PR )

File: web/client/plugins/Swipe.jsx (hunk 16 above)
Comment (allyoucanmap) [reply web/client/plugins/Swipe.jsx:25]: @offtherailz here the availability of plugin it's seem solved with disablePluginIf, (imo) if there is still work to do on this Support implementation we could improve this later on a new PR.

**Review by offtherailz — APPROVED** [review]
